File: test_c/test_fibmem.c
#include <stdio.h>
#include <stdlib.h>

long long fibmem(int n, long long *memo, int *opC) {
    if (n <= 2) { memo[n] = 1; return 1; }
    if (memo[n] != 0) { return memo[n]; }
    memo[n] = fibmem(n-1, memo, opC) + fibmem(n-2, memo, opC);
    (*opC)++;
    return memo[n];
}

int main() {
    int n = 93;
    long long *memo = calloc(n + 1, sizeof(long long));
    int opC = 0;
    int passed = 0;
    if (fibmem(1, memo, &opC) == 1)                        { printf("test 1 passed\n"); passed++; }
    if (fibmem(2, memo, &opC) == 1)                        { printf("test 2 passed\n"); passed++; }
    if (fibmem(7, memo, &opC) == 13)                       { printf("test 3 passed\n"); passed++; }
    if (fibmem(10, memo, &opC) == 55)                      { printf("test 4 passed\n"); passed++; }
    if (fibmem(20, memo, &opC) == 6765)                    { printf("test 5 passed\n"); passed++; }
    if (fibmem(93, memo, &opC) == 12200160415121876738ULL) { printf("test 6 passed\n"); passed++; }
    free(memo);
    printf("%d/6 tests passed.\n", passed);
    return 0;
}

File: test_rs/test_fibmem.rs
use std::collections::HashMap;

fn fibmem(n: u64, memo: &mut HashMap<u64, u64>, op_c: &mut u64) -> u64 {
    if n <= 2 { return 1; }
    if let Some(val) = memo.get(&n) { return *val; }
    let current = fibmem(n-1, memo, op_c) + fibmem(n-2, memo, op_c);
    *op_c += 1;
    memo.insert(n, current);
    current
}

fn main() {
    let mut memo = HashMap::new();
    let mut op_c: u64 = 0;
    let mut passed = 0;
    if fibmem(1, &mut memo, &mut op_c) == 1                     { println!("test 1 passed"); passed += 1; }
    if fibmem(2, &mut memo, &mut op_c) == 1                     { println!("test 2 passed"); passed += 1; }
    if fibmem(7, &mut memo, &mut op_c) == 13                    { println!("test 3 passed"); passed += 1; }
    if fibmem(10, &mut memo, &mut op_c) == 55                   { println!("test 4 passed"); passed += 1; }
    if fibmem(20, &mut memo, &mut op_c) == 6765                 { println!("test 5 passed"); passed += 1; }
    if fibmem(93, &mut memo, &mut op_c) == 12200160415121876738 { println!("test 6 passed"); passed += 1; }
    println!("{}/6 tests passed.", passed);
}
